localstack/services/secretsmanager/provider.py

```diff
@@ -0,0 +1,261 @@
+import json
+import logging
+
+from localstack.aws.api import HttpRequest, RequestContext
+from localstack.aws.api.secretsmanager import (
+    AddReplicaRegionListType,
+    BooleanType,
+    CancelRotateSecretResponse,
+    ClientRequestTokenType,
+    CreateSecretResponse,
+    DeleteResourcePolicyResponse,
+    DeleteSecretResponse,
+    DescribeSecretResponse,
+    DescriptionType,
+    GetResourcePolicyResponse,
+    GetSecretValueResponse,
+    KmsKeyIdType,
+    ListSecretVersionIdsResponse,
+    MaxResultsType,
+    NameType,
+    NextTokenType,
+    NonEmptyResourcePolicyType,
+    PutResourcePolicyResponse,
+    PutSecretValueResponse,
+    RecoveryWindowInDaysType,
+    RemoveRegionsFromReplicationResponse,
+    RemoveReplicaRegionListType,
+    ReplicateSecretToRegionsResponse,
+    RestoreSecretResponse,
+    RotateSecretResponse,
+    RotationLambdaARNType,
+    RotationRulesType,
+    SecretBinaryType,
+    SecretIdType,
+    SecretsmanagerApi,
+    SecretStringType,
+    SecretVersionIdType,
+    SecretVersionStagesType,
+    SecretVersionStageType,
+    StopReplicationToReplicaResponse,
+    TagKeyListType,
+    TagListType,
+    UpdateSecretResponse,
+    UpdateSecretVersionStageResponse,
+    ValidateResourcePolicyResponse,
+)
+from localstack.services.moto import call_moto
+from localstack.services.secretsmanager.secretsmanager_patches import apply_patches
+from localstack.utils.aws import aws_stack
+from localstack.utils.common import to_str
+
+LOG = logging.getLogger(__name__)
+
+
+class SecretsmanagerProvider(SecretsmanagerApi):
+    def __init__(self):
+        apply_patches()  # Adds missing moto attributes.
+
+    @staticmethod
+    def __transform_context_secret_id(context: RequestContext) -> RequestContext:
+        data_dict = json.loads(to_str(context.request.data or "{}"))
+        secret_id = data_dict.get("SecretId", "")
+        if ":" in secret_id:
+            request = context.request
+            parts = secret_id.split(":")
+            if parts[3] != aws_stack.get_region():
+                LOG.info(
+                    'Unexpected request region %s for secret "%s"',
+                    aws_stack.get_region(),
+                    secret_id,
+                )
+            # secret ARN ends with "-<randomId>" which we remove in the request for upstream compatibility
+            # if the full arn is being sent then we remove the string in the end
+            if parts[-1][-7] == "-":
+                data_dict["SecretId"] = parts[-1][: len(parts[-1]) - 7]
+            elif parts[-1][-1] != "-":
+                data_dict["SecretId"] = data_dict["SecretId"] + "-"
+            #
+            context.request = HttpRequest(
+                method=request.method,
+                path=request.path,
+                query_string=request.query_string,
+                headers=request.headers,
+                body=bytes(json.dumps(data_dict), "utf-8"),
```

Comment:
nit: better use the `to_bytes(..)` util function from `common.py` (which internally uses the default encoding configured):
```
body=to_bytes(json.dumps(data_dict))
```
(but this will anyway be obsolete if we use `call_moto_with_request(..)`, as outlined above)